types/underscore/index.d.ts

```diff
@@ -856,10 +849,13 @@ declare module _ {
          * corresponding to those keys.
          * @retursn An object comprised of the provided keys and values.
          **/
-        object<V extends List<any>, TValueList extends List<any> | undefined = undefined>(
-            list: V,
-            values?: TValueList
-        ): Dictionary<_ObjectValue<TypeOfList<V>, TValueList>>;
+        object<TList extends List<any>, TValue>(
```

Comment:
Isn't `any` way too liberal here?

```suggestion
        object<TList extends List<EnumerableKey>, TValue>(
```

Reply:
Most of the time I would expect folks to be providing keys in the form of strings or numbers, and it is definitely not great for someone to not do that and end up with an object that has a single "[object Object]" property, but it is technically valid to use anything that provides a string representation of itself, and I wasn't sure if it was worth blocking that since it can work and since the existing definitions were allowing it.

Two examples of weird inputs that technically provide valid and useful outputs are below. I probably wouldn't ever do the second one, but I admittedly don't entirely hate the first one.

Using an array to come up with a dictionary that allows one to look up values based on multiple keys (also something one could and is probably more likely to do with `indexBy` which is why I also chose to leave that unconstrained in #46352):
`_.object([['one', 1], ['two', 2]], ['blah', 'foo'])` => `{"one,1": "blah", "two,2": "foo"}`
```TS
const keys = [[1, 'a'], [1, 'b'], [2, 'a'], [2, 'b']];
const values = ['1a Value', '1b Value', '2a Value', '2b Value'];
const rowColumnLookup = _.object(keys, values);

declare const row: number;
declare const column: string;
const itemAtColumnAndRow = rowColumnLookup[[row, column]];
```

Making up some sort of extended string class for reasons:
`_.object([{ [Symbol.toPrimitive]: () => 'abrac' }, { [Symbol.toPrimitive]: () => 'adabra' }], ['blah', 'foo'])` => `{"abrac": "blah", "adabra": "foo"}`
```TS
class PrefixedString {
    private static readonly prefix = 'pre-';
    
    public constructor(public str: string) {}

    [Symbol.toPrimitive]: () => PrefixedString.prefix + this.str;
}

const keys = [new PrefixedString('one'), new PrefixedString('two')];
const values = [1, 2];
const prefixedStringObject = _.object(keys, values);

declare const prefixed: PrefixedString;
const itemFromPrefixedString = prefixedStringObject[prefixed];
```

Reply:
If you'd prefer to provide some guard rails here, I'd also be fine with making folks `toString` any oddness they choose to run through this to help folks that might otherwise accidentally do unfortunate things. I don't have a strong desire for this to be this loose for any code I might have to mess with.

Reply:
Ha, all that stuff works in JS but I actually tried it in TS just now and TS is for sure against the idea of doing `const itemAtColumnAndRow = rowColumnLookup[[row, column]];` (and probably the lookup by `prefixed` as well) 😬. Never mind, I'll add the constraint.

Reply:
As a final point of interest here, I did just realize that the return type of object constrains its indexing to `string` specifically, which is something that I can't get around because index declarations can't use `EnumerableKey` or even `any`. However, it seems like JS probably coerces numbers to strings when indexing into objects (`({ "0": "a" })[0]` => `"a"`) and it looks like TS recognizes that since it still allows numbers to be used as index values even on objects with a declared index type of `string`. However, even `anyValue[[1, 2]];` where `declare const anyValue: any` doesn't work, so I still think this is the right thing to do to be consistent with TS's strong views on allowed indexing types since folks can `toString` or `Symbol.toPrimitive` non-strings themselves both when providing index values for objects and when indexing into those objects.